docs/api-docs.rst

```diff
@@ -1,6 +1,217 @@
+        (, "inputs": [<TASK_INPUT> (, <TASK_INPUT>, ...)])
+        (, "outputs": [<TASK_OUTPUT> (, <TASK_OUTPUT>, ...)])
+    }
+
+    <R_TASK> ::= {
+        "mode": "r",
+        "script": <r code to run (as a string)>
+        (, "inputs": [<TASK_INPUT> (, <TASK_INPUT>, ...)])
+        (, "outputs": [<TASK_OUTPUT> (, <TASK_OUTPUT>, ...)])
+    }
+
+    <DOCKER_TASK> ::= {
+        "mode": "docker",
+        "docker_image": <docker image name to run>
+        (, "container_args": [<container arguments>])
+        (, "inputs": [<TASK_INPUT> (, <TASK_INPUT>, ...)])
+        (, "outputs": [<TASK_OUTPUT> (, <TASK_OUTPUT>, ...)])
+    }
+
+    <WORKFLOW_TASK> ::= {
+        "mode": "workflow",
+        "steps": [<WORKFLOW_STEP> (, <WORKFLOW_STEP>, ...)],
+        "connections": [<WORKFLOW_CONNECTION> (, <WORKFLOW_CONNECTION>, ...)]
+        (, "inputs": [<TASK_INPUT> (, <TASK_INPUT>, ...)])
+        (, "outputs": [<TASK_OUTPUT> (, <TASK_OUTPUT>, ...)])
+    }
+
+    <WORKFLOW_STEP> ::= {
+        "name": <step name>,
+        "task": <TASK>
+    }
+
+    <WORKFLOW_CONNECTION> ::= {
+        ("name": <name of top-level input to bind to>)
+        (, "input": <input id to bind to for a step>)
+        (, "input_step": <input step name to connect>)
+        (, "output_step": <output step name to connect>)
+    }
+
+The workflow mode simply allows for a directed acyclic graph of tasks to be
+specified to :py:func:`romanesco.run`.
+
+.. code-block:: none
+
+    <TASK_INPUT> ::= {
+        "id": <string, the variable name>,
+        "type": <data type>,
+        "format": <data format>
+        (, "target": <INPUT_TARGET_TYPE>)   ; default is "memory"
+        (, "filename": <name of file if target="filepath">)
+    }
+
+    <INPUT_TARGET_TYPE> ::= "memory" | "filepath"
+
+    <TASK_OUTPUT> ::= {
+        "id": <string, the variable name>,
+        "type": <data type>,
+        "format": <data format>
+        (, "target": <INPUT_TARGET_TYPE>)   ; default is "memory"
+    }
+
+
+The input specification
+***********************
+
+The ``inputs`` argument to :py:func:`romanesco.run` specifies the inputs to the
+task described by the ``task`` argument. Specifically, it tells what data should
+be placed into the task input ports.
+
+.. code-block:: none
+
+    <INPUTS> ::= {
+        <id> : <INPUT_BINDING>
+    }
+
+The input spec is a dictionary mapping each ``id`` (corresponding to the ``id`` key of
+each task input) to its data binding for this execution.
+
+.. code-block:: none
+
+    <INPUT_BINDING> ::= <INPUT_BINDING_HTTP> | <INPUT_BINDING_LOCAL> |
+                        <INPUT_BINDING_MONGODB> | <INPUT_BINDING_INLINE>
+
+    <INPUT_BINDING_HTTP> ::= {
+        "mode": "http",
+        "url": <url of data to download>
+        (, "headers": <dict of HTTP headers to send when fetching>)
+        (, "method": <http method to use, default is "GET">)
+        (, "maxSize": <integer, max size of download in bytes>)
+    }
```

Comment:
I believe all input bindings should have formats. I see the `format` field for output bindings but not input bindings.